mne/preprocessing/faster.py

```diff
@@ -0,0 +1,368 @@
+        'power_gradient':  lambda x: _power_gradient(source_data,
+                                                     x.info['sfreq']),
+        'hurst':           lambda x: _hurst(source_data),
+        'median_gradient': lambda x: np.median(np.abs(np.diff(source_data)),
+                                               axis=1),
+        'line_noise':      lambda x: _freqs_power(source_data,
+                                                  epochs.info['sfreq'],
+                                                  [50, 60]),
+    }
+
+    if use_metrics is None:
+        use_metrics = metrics.keys()
+
+    bads = []
+    for m in use_metrics:
+        scores = np.atleast_2d(metrics[m](ica))
+        for s in scores:
+            b = find_outliers(s, thres)
+            logger.info('Bad by %s:\n\t%s' % (m, b))
+            bads.append(b)
+
+    return np.unique(np.concatenate(bads)).tolist()
+
+
+def faster_bad_channels_in_epochs(epochs, picks=None, thres=3,
+                                  use_metrics=None):
+    """Implements the fourth step of the FASTER algorithm.
+    
+    This function attempts to automatically mark bad channels in each epochs by
+    performing outlier detection.
+
+    Parameters
+    ----------
+    epochs : Instance of Epochs
+        The epochs to analyze.
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
+    thres : float
+        The threshold value, in standard deviations, to apply. An epoch
+        crossing this threshold value is marked as bad. Defaults to 3.
+    use_metrics : list of str
+        List of metrics to use. Can be any combination of:
+            'amplitude', 'variance', 'deviation', 'median_gradient'
+        Defaults to all of them.
+
+    Returns
+    -------
+    bads : list of lists of int
+        For each epoch, the indices of the bad channels.
+    """
+
+    metrics = {
+        'amplitude':       lambda x: np.ptp(x, axis=2),
+        'deviation':       lambda x: _deviation(x),
+        'variance':        lambda x: np.var(x, axis=2),
+        'median_gradient': lambda x: np.median(np.abs(np.diff(x)), axis=2),
+        'line_noise':      lambda x: _freqs_power(x, epochs.info['sfreq'],
+                                                  [50, 60]),
+    }
+
+    if picks is None:
+        picks = pick_types(epochs.info, meg=False, eeg=True, exclude='bads')
+    if use_metrics is None:
+        use_metrics = metrics.keys()
+    
+    data = epochs.get_data()[:, picks, :]
+
+    bads = [[] for i in range(len(epochs))]
+    for m in use_metrics:
+        s_epochs = metrics[m](data)
+        for i, s in enumerate(s_epochs):
+            b = [epochs.ch_names[picks[j]] for j in find_outliers(s, thres)]
+            if len(b) > 0:
+                logger.info('Epoch %d, Bad by %s:\n\t%s' % (i, m, b))
+            bads[i].append(b)
+
+    for i, b in enumerate(bads):
+        if len(b) > 0:
+            bads[i] = np.unique(np.concatenate(b)).tolist()
```

Comment:
looks like a filter list comprehension could do it much easier:

`bads =  [np.unique(np.concatenate(b)).tolist() for b in bads if len(b) >0]`